controlpanel/api/tools.py

```diff
@@ -0,0 +1,212 @@
+import secrets
+
+from django.conf import settings
+from django.utils.functional import cached_property
+from rest_framework.exceptions import APIException
+
+from controlpanel.api.helm import Helm
+from controlpanel.kubeapi.views import client
+
+
+IDLED = "mojanalytics.xyz/idled"
+
+SUPPORTED_TOOL_NAMES = list(settings.TOOLS.keys())
+
+
+class ToolNotDeployed(APIException):
+    status_code = 400
+    default_detail = f"Specified tool not deployed"
+    default_code = "tool_not_deployed"
+
+
+class UnsupportedToolException(APIException):
+    status_code = 400
+    supported = ", ".join(SUPPORTED_TOOL_NAMES)
+    default_detail = f"Unsupported tool, must be one of: {supported}"
+    default_code = "unsupported_tool"
+
+
+tools_repository = dict()
+
+
+class ToolMeta(type):
+    """Adds Tool subclasses to tools_repository"""
+
+    def __init__(cls, name, bases, clsdict):
+        super().__init__(name, bases, clsdict)
+        name = clsdict["name"]
+        tools_repository[name] = cls
+
+
+class Tool(metaclass=ToolMeta):
+    name = None
+
+    @classmethod
+    def create(cls, name):
+        try:
+            return tools_repository[name]()
+        except KeyError:
+            raise UnsupportedToolException()
```

Comment:
#### 1e. Refactor Tools code
> Lines 41 - 130
I refactored the ToolRepository class into a simple dict, with entries automatically added by the `ToolMeta` class when a `Tool` subclass is defined. *RETHINK THIS*